pkg/object/mqttproxy/broker.go

```diff
@@ -450,6 +553,87 @@ func (b *Broker) sendMsgToClient(span *model.SpanContext, topic string, payload
 	}
 }
 
+// splitSubscribers split subscribers to local and remote on broker mode and return
+// client ids of local subscribers and eg names of remote subscribers
+func (b *Broker) splitSubscribers(span *model.SpanContext, publish *packets.PublishPacket) ([]string, map[string]struct{}) {
+	egNames := make(map[string]struct{})
+	clients := []string{}
+
+	subscribers, _ := b.topicMgr.findSubscribers(publish.TopicName)
+	for clientID, subQos := range subscribers {
+		if subQos < publish.Qos {
+			continue
+		}
+		egName := b.sessionCacheMgr.getEGName(clientID)
+		if egName != b.egName {
+			egNames[egName] = struct{}{}
+			continue
+		}
+		clients = append(clients, clientID)
+	}
+	return clients, egNames
+}
+
+func (b *Broker) sendMsgToLocalClient(span *model.SpanContext, publish *packets.PublishPacket, clients []string) {
+	for _, clientID := range clients {
+		client := b.getClient(clientID)
+		if client == nil {
+			logger.SpanDebugf(span, "client %v not on broker %v in eg %v", clientID, b.name, b.egName)
+		} else {
+			client.session.publish(span, publish.TopicName, publish.Payload, publish.Qos)
+		}
+	}
+}
+
+func (b *Broker) requestTransferToCertainInstances(span *model.SpanContext, remoteEgs map[string]struct{}, publish *packets.PublishPacket) {
+	data := &HTTPJsonData{}
+	data.init(publish)
+	urls, err := b.memberURL(b.egName, b.name)
+	if err != nil {
+		logger.SpanErrorf(span, "eg %v find urls for other egs failed: %v", b.egName, err)
+		return
+	}
+	jsonData, err := codectool.MarshalJSON(data)
+	if err != nil {
+		logger.SpanErrorf(span, "json data marshal failed: %v", err)
+		return
+	}
+	for egName := range remoteEgs {
+		url, ok := urls[egName]
+		if !ok {
+			logger.SpanErrorf(span, "eg %s not find url for eg %s", b.egName, egName)
+			continue
+		}
+		req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(jsonData))
+		if err != nil {
+			logger.SpanErrorf(span, "make new request failed: %v", err)
+			continue
+		}
+		err = b3.InjectHTTP(req, b3.WithSingleHeaderOnly())(*span)
+		if err != nil {
+			logger.SpanErrorf(span, "inject span failed: %v", err)
+			continue
+		}
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			logger.SpanErrorf(span, "http client send msg failed:%v", err)
+		} else {
+			resp.Body.Close()
+		}
+	}
+
+}
+
+func (b *Broker) processBrokerModePublish(clientID string, publish *packets.PublishPacket) {
+	if !b.spec.BrokerMode {
+		return
+	}
+	span := generateNewSpanContext(clientID, publish.TopicName)
+	localClients, remoteEgs := b.splitSubscribers(span, publish)
```

Comment:
0. In the function splitSubscribers, the argument span is redundant.
1. It is possible to call the function `generateNewSpanContext` after the call to function `splitSubscribers`.
2. After calling of `splitSubscribers`, we need to check whether the variables of the`localClients` and the `remoteEgs` are both empty. If they are empty, we needn't to call the function `sendMsgToLocalClient` and the function `requestTransferToCertainInstances`
3.  The arguments of `publish` to the functions `sendMsgToLocalClient` and `requestTransferToCertainInstances` should have the same position. The current position of the argument is counter-intuitive.

Reply:
thanks a lot for your advice, I will update it now.

Reply:
updated!